### DISCUSSION TITLE
How to pass ' –pretty=fuller' option to git log command in opengrok index process

### DISCUSSION THREAD

**Discussion by sreehari-malapati-myprod** [oracle/opengrok#4052] [Q&A]
How to pass ' –pretty=fuller' option to git log command in opengrok index process

**Comment by vladak** [comment]
Why exactly do you need that ? Git handling within OpenGrok indexer has been converted to JGit so git command line options are no longer relevant.

**Comment by sreehari-malapati-myprod** [comment]
By default git log is giving author date as commit date  and author date is showing in history tab of opengrok UI.  
git log is giving correct commit date if we use –pretty=fuller option in git log

**Reply by vladak** [reply]
Okay, I [see](https://stackoverflow.com/a/11857467/11582827) [what](https://stackoverflow.com/a/69215450/11582827) you mean:
```
$ git log -1 4fe126c3e66f6654aca414b1df4557f1c122572a
commit 4fe126c3e66f6654aca414b1df4557f1c122572a
Author: Vladimír Kotal <vlada@kotalovi.cz>
Date:   Thu Oct 6 23:23:27 2022 +0200

    move PeriodicTimer class to separate file

$ git log --pretty=fuller -1 4fe126c3e66f6654aca414b1df4557f1c122572a
commit 4fe126c3e66f6654aca414b1df4557f1c122572a
Author:     Vladimír Kotal <vlada@kotalovi.cz>
AuthorDate: Thu Oct 6 23:23:27 2022 +0200
Commit:     Vladimir Kotal <vlada@kotalovi.cz>
CommitDate: Fri Oct 7 09:21:40 2022 +0200

    move PeriodicTimer class to separate file

```
The question is which of these dates is the "correct" one.

Currently, the `AuthorDate` is used for the history entries and `CommitDate` is used to get the current version metadata for a repository (`GitRepository#determineCurrentVersion()`) that is displayed on the index page in project/repository listing and also when building tag list (`GitRepository#rebuildTagList()`).

Now, `HistoryEntry#date` (that is used for showing the date in history view in the UI) for Git repositories has been always based on `AuthorDate` (which is probably the reason I chose `RevCommit#getAuthorIdent().getWhen()` when rewriting `GitRepository` to `JGit` - so that there is no change in behavior). Even if the `CommitDate` is chosen as the "correct" one, that would mean recreating the history cache from scratch so that it stores consistent dates.

Github in their UI appears to be showing the `CommitDate` even though the API can get both dates:
```
$ curl -s https://api.github.com/repos/oracle/opengrok/commits?path=docker/start.py | jq ".[0].commit"
{
  "author": {
    "name": "Vladimír Kotal",
    "email": "vlada@kotalovi.cz",
    "date": "2022-10-06T21:23:27Z"
  },
  "committer": {
    "name": "Vladimir Kotal",
    "email": "vlada@kotalovi.cz",
    "date": "2022-10-07T07:21:40Z"
  },
  "message": "move PeriodicTimer class to separate file",
  "tree": {
    "sha": "bce9f2f9317d949a66ec654d7a726515ea689a18",
    "url": "https://api.github.com/repos/oracle/opengrok/git/trees/bce9f2f9317d949a66ec654d7a726515ea689a18"
  },
  "url": "https://api.github.com/repos/oracle/opengrok/git/commits/4fe126c3e66f6654aca414b1df4557f1c122572a",
  "comment_count": 0,
  "verification": {
    "verified": false,
    "reason": "unsigned",
    "signature": null,
    "payload": null
  }
}
```

I can be convinced either way so feel free to start what seems to be a philosophical dispute. The main question is: why should commit date be preferred over author date ? Some say that it is more convenient for [resolving a merge conflict](https://github.com/gitkraken/vscode-gitlens/issues/537#issuecomment-441921271).

Also, having common date to be used throughout the UI is probably beneficial and the current situation (2 x commitDate vs 1 x authorDate) would speak for commitDate.

**Comment by maestropandy** [comment]
@vladak I have issues somehow related to this, In my case, Opengrok UI is displaying date as per in the git log, but in git log we have option to sort date using below command

_git log --pretty=fuller --follow --date-order <filename>_

Whether Opengrok can sort date order as above git command  does ?

**Reply by vladak** [reply]
No, the history strictly follows changesets.

**Reply by vladak** [reply]
> No, the history strictly follows changesets.

That said, it might be possible to implement the sorting in the UI, similarly to how the sorting is done in directory listings.

**Reply by vladak** [reply]
> > No, the history strictly follows changesets.
> 
> That said, it might be possible to implement the sorting in the UI, similarly to how the sorting is done in directory listings.

Thinking if this a little bit more, this might pose some challenges - esp. the comparison of changesets - when the order is changed in the UI, it would be possible to get a diff between two changesets however conceptually it would be skewed as the changesets in between might be out of order w.r.t. the real changeset ordering in the repository.

**Reply by maestropandy** [reply]
@vladak I have revisited file change, where git log <filename> shows "date" order different than in Opengrok UI. In our case Authordate and Commit date are same for that file, still we are seeing date in Opengrok as result of  `git log --pretty=fuller --follow`, we are looking opengrok to sort date of file as result "git log <filename>"
